src/platform/cmx/chal.c

```diff
@@ -0,0 +1,206 @@
+		}
+	}
+	else {
+		for(count=0;count<8;count++) {
+			if(((table->data.pgt[count]&CMX_PGTBL_PRESENT)!=0)&&((table->data.pgt[count]&CMX_PGTBL_LEAF)!=0))
+				rasr|=(((u32_t)1)<<(count+8));
+		}
+	}
+
+	return rasr;
+}
+
+
+/* Generate the RASR data */
+u32_t
+chal_mpu_gen_rasr(struct pgtbl* table, u32_t size_order, u32_t num_order)
+{
+	u32_t rasr;
+
+	/* Get the SRD part first */
+	rasr=chal_mpu_gen_srd(table, num_order);
+	if(rasr==0)
+		return 0;
+
+	rasr=CMX_MPU_SRDCLR&(~rasr);
+	rasr|=CMX_MPU_SZENABLE;
+	/* Because what we type is always read-write, thus the RASR is always RW */
+	rasr|=CMX_MPU_RW;
+	/* Can we fetch instructions from there? - always cannot, we do not allow data execution */
+	rasr|=CMX_MPU_XN;
+	/* Is the area cacheable? - because what we type is SRAM, so it is always accessible */
+	rasr|=CMX_MPU_CACHEABLE;
+	/* Is the area bufferable? - always bufferable */
+	rasr|=CMX_MPU_BUFFERABLE;
+	/* What is the region size? */
+	rasr|=CMX_MPU_REGIONSIZE(size_order, num_order);
+        /* 1303fc1b */
+	return rasr;
+}
+
+u32_t
+chal_mpu_clear(struct pgtbl* table, u32_t start_addr, u32_t size_order)
+{
+	u32_t count;
+
+	/* There are only 4 regions at the toplevel which is configurable */
+	for(count=0; count<4; count++) {
+		if((table->data.pgt[9] & (((u32_t)1)<<count)) != 0) {
+			/* We got one MPU region valid here */
+			if((CMX_MPU_ADDR(table->data.mpu[count].rbar)==start_addr)&&
+			   (CMX_MPU_SZ_ORD(table->data.mpu[count].rasr)==size_order)) {
+				/* Clean it up and return */
+				table->data.mpu[count].rbar = CMX_MPU_VALID|count;
+				table->data.mpu[count].rasr = 0;
+				table->data.pgt[9] &= ~(((u32_t)1) << count);
+				return 0;
+			}
+		}
+	}
+
+	return 0;
+}
+
+u32_t
+chal_mpu_add(struct pgtbl* table, u32_t start_addr, u32_t size_order, u32_t mpu_rasr)
+{
+	u32_t count;
+	/* The last empty slot */
+	u32_t last_empty;
+
+	/* Set these values to some overrange value */
+	last_empty=16;
+
+	/* We only set 4 regions */
+	for(count=0;count<4;count++) {
+		if((table->data.pgt[9]&(((u32_t)1)<<count))!=0) {
+			/* We got one MPU region valid here */
+			if((CMX_MPU_ADDR(table->data.mpu[count].rbar)==start_addr)&&   /* RBAR */
+			   (CMX_MPU_SZ_ORD(table->data.mpu[count].rasr)==size_order)) /* RASR */ {
+				/* Update the RASR - all flag changes except static are reflected here */
+				table->data.mpu[count].rasr=mpu_rasr;
+				return 0;
+			}
+		}
+		else
+			last_empty=count;
+	}
+
+	/*
+	 * Update unsuccessful, we didn't find any match. We will need a new slot.
+	 * See if there are any new empty slots that we can use.
+	 * See if the last empty slot is 0. If yes, we can only map dynamic pages
+	 */
+	if(last_empty!=16) {
+		/* Put the data to this slot */
+		table->data.pgt[9]|=((u32_t)1)<<(last_empty);
+		table->data.mpu[last_empty].rbar=CMX_MPU_ADDR(start_addr)|CMX_MPU_VALID|last_empty;
+		table->data.mpu[last_empty].rasr=mpu_rasr;
+
+		return 0;
+	}
+
+	/* All effort is futile. we report failure */
+	return -1;
```

Comment:
u32_t is the return type. 
It's odd to have `-1` as return value here. You should probably change the return type to `signed`.